src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
+              workerPool.workerDone(worker);
             }
           }
         } catch (InterruptedException e) {
+          if (worker != null) {
+            workerPool.workerDone(worker);
+          }
           stopIfRunning();
           Thread.currentThread().interrupt();
         } catch (Throwable t) {
+          if (worker != null) {
+            // we kill the worker here because this condition is not expected
+            workerPool.killWorker(worker);
+          }
           ExceptionUtils.handleException(t);
+        } finally {
+          worker = null;
         }
       }
     }
     
-    protected TaskWrapper getNextTask() throws InterruptedException {
+    /**
+     * Gets the next task that is ready to execute immediately.  This call gets the next high and 
+     * low priority task, compares them and determines which one should be executed next.
+     * 
+     * If there are no tasks ready to be executed this will block until one becomes available 
+     * and ready to execute.
+     * 
+     * @return Task to be executed next, or {@code null} if shutdown while waiting for task
+     * @throws InterruptedException Thrown if calling thread is interrupted while waiting for a task
+     */
+    protected TaskWrapper getNextReadyTask() throws InterruptedException {
       while (runningThread != null) {  // loop till we have something to return
-        TaskWrapper nextScheduledTask = scheduleQueue.peekFirst();
-        TaskWrapper nextExecuteTask = executeQueue.peek();
-        if (nextExecuteTask != null) {
-          if (nextScheduledTask != null) {
-            long scheduleDelay;
-            long executeDelay;
-            ClockWrapper.stopForcingUpdate();
-            try {
-              scheduleDelay = nextScheduledTask.getDelay(TimeUnit.MILLISECONDS);
-              executeDelay = nextExecuteTask.getDelay(TimeUnit.MILLISECONDS);
-            } finally {
-              ClockWrapper.resumeForcingUpdate();
-            }
-            if (scheduleDelay < executeDelay) {
-              synchronized (scheduleQueue.getModificationLock()) {
-                // scheduled tasks must be removed, and call .executing() while holding the lock
-                if (scheduleQueue.remove(nextScheduledTask)) {
-                  nextScheduledTask.executing();
-                  return nextScheduledTask;
-                }
+        TaskWrapper nextHighPriorityTask = highPriorityQueueSet.getNextTask();
+        TaskWrapper nextLowPriorityTask = lowPriorityQueueSet.getNextTask();
+        if (nextLowPriorityTask == null) {
+          if (nextHighPriorityTask == null) {
+            // no tasks, so just wait till any tasks come in
+            LockSupport.park();
+          } else {
+            // only high priority task is queued, so run if ready
+            long nextTaskDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            if (nextTaskDelay <= 0) {
+              if (nextHighPriorityTask.canExecute()) {
```

Comment:
It actually is in a while loop.  If you look at line 1187 (new line number), you will see we loop till we find a task to return, or until we shutdown the scheduler.

You are right, tasks can be removed (not because of executing, but because of a manual removal).  This is where we make sure we can automically remove the task for execution.